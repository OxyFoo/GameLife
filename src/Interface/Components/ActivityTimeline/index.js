import * as React from 'react';
import { View, FlatList } from 'react-native';

import styles from './style';
import ActivityTimelineBack from './back';

/**
 * @typedef {import('./back').ActivityTimelineItem} ActivityTimelineItem
 * @typedef {import('react-native').ListRenderItem<ActivityTimelineItem>} ListRenderItemActivityTimelineItem
 */

class ActivityTimeline extends ActivityTimelineBack {
    render() {
        const { activities } = this.state;

        return (
            <View style={[styles.parent, this.props.style]} onLayout={this.onLayout}>
                <FlatList
                    data={activities}
                    renderItem={this.renderActivity}
                    keyExtractor={(item, index) => `activity-${item.startTime}-${index}`}
                    showsHorizontalScrollIndicator={false}
                    scrollEnabled={false}
                    horizontal={true}
                />
            </View>
        );
    }

    /** @type {ListRenderItemActivityTimelineItem} */
    renderActivity = ({ item }) => {
        return (
            <View
                style={[
                    styles.timelineItem,
                    {
                        marginLeft: item.marginLeft,
                        width: item.width,
                        borderColor: item.color
                    },
                    item.hasPreviousAdjacentActivity && styles.adjacentLeft,
                    item.hasNextAdjacentActivity && styles.adjacentRight
                ]}
            />
        );
    };
}

export { ActivityTimeline };
