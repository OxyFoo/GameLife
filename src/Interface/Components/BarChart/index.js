import React, { useEffect, useState } from 'react';
import { View, Image, Text } from 'react-native';
import { LineChart, BarChart } from "react-native-gifted-charts"

import styles from './style';
import BarChartBack from './back';

class BarChartSkill extends BarChartBack {
    render() {

        return (
            <View
                style={styles.container}>
                {this.dataReady && this.state.cleanedData ?
                    <BarChart
                        data={this.state.cleanedData}
                        width={this.props.chartWidth}
                        barWidth={this.barWidth}
                        initialSpacing={0}
                        spacing={0}
                        cappedBars
                        capColor={'rgba(78, 0, 142)'}
                        capThickness={4}
                        showGradient
                        gradientColor={'rgba(200, 100, 244,0.8)'}
                        frontColor={'rgba(219, 182, 249,0.2)'}
                        yAxisTextStyle={{color: 'white'}}
                        yAxisColor={'white'}
                        xAxisColor={'white'}
                        isAnimated
                    />
                    : <></>
                }
            </View>
        )
    }
};

export default BarChartSkill; 