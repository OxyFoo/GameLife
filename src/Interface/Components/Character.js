import * as React from 'react';
import { Animated, View, StyleSheet } from 'react-native';
import Svg from 'react-native-svg';

import Character_Classic from '../../../res/stuffs/humans/human_classic';
import Character_Male_Test from '../../../res/stuffs/humans/human_male_test';

import { TimingAnimation, WithFunction } from '../../Utils/Animations';

class Part {
    /**
     * @param {String} name Name of body part
     * @param {Number} l Distance from parent in pixels (for initial position)
     * @param {Number} a Angle distance from parent in degrees (for initial position)
     * @param {Animated.Value} rotation Rotation of body part in degrees
     * @param {Number} [zIndex=0] Z-index of body part
     */
    constructor(name, l, a, rotation, zIndex = 0) {
        this.name = name;
        this.zIndex = zIndex;

        // Initial position from parent
        this.l = l;
        this.a = a;
        //this.r = 0;

        // Current absolute position
        this.rotation = rotation;
        this.pos = new Animated.ValueXY({ x: 0, y: 0 });

        /** @type {Part} */
        this.parent = null;

        /** @type {Array<Part>} part */
        this.childs = [];
    }

    /**
     * @param {Animated.ValueXY} position Position of body part in pixels (x, y)
     */
    SetPosition(position) {
        this.pos = position;
    }

    /**
     * @param {Part} child 
     */
    AddChild(child) {
        child.parent = this;
        this.childs.push(child);
    }
    /**
     * @param {String} name
     * @returns {Part?}
     */
    GetChild(name) {
        return this.childs.find(part => part.name === name) || null;
    }

    /**
     * @description Return an array of all parent parts in order from root to this part (including this part)
     * @returns {Array<Part>}
     */
    getParents() {
        let parents = [ this ];
        let parent = this.parent;
        while (parent !== null) {
            parents.splice(0, 0, parent);
            parent = parent.parent;
        }
        return parents;
    }

    render() {
        //const TETA = this.a * Math.PI / 180;
        //let posX = Animated.add(this.pos.x, this.l * Math.cos(TETA));
        //let posY = Animated.add(this.pos.y, this.l * Math.sin(TETA));
        //let rot = this.rotation;
        //console.log(this.name, this.pos, this.rotation);

        /*if (this.parent !== null) {
            const TETA = (this.parent.r + this.a) * Math.PI / 180;
            posX = Animated.add(posX, this.l * Math.cos(TETA));
            posY = Animated.add(posY, this.l * Math.sin(TETA));
            //this.r += this.parent.r;
        }*/

        /*let el = this;
        while (el.parent !== null) {
            let totalRotation = Animated.add(el.a, el.parent.rotation);
            totalRotation = Animated.add(totalRotation, el.parent.a);
            const TETA = Animated.multiply(totalRotation, Math.PI / 180);

            const COS = TETA.interpolate(WithFunction(Math.cos, 20, 2*Math.PI));
            const SIN = TETA.interpolate(WithFunction(Math.sin, 20, 2*Math.PI));
            const radiusX = Animated.multiply(this.l, COS);
            const radiusY = Animated.multiply(this.l, SIN);

            posX = Animated.add(posX, radiusX);
            posY = Animated.add(posY, radiusY);
            posX = Animated.add(posX, el.parent.pos.x);
            posY = Animated.add(posY, el.parent.pos.y);

            const parentRotation = Animated.add(el.parent.a, el.parent.rotation);
            rot = Animated.add(rot, el.r);
            rot = Animated.add(rot, parentRotation);
            el = el.parent;
        }*/
        /*const calculateParentPos = (el) => {
            if (el.parent !== null) {
                const totalRotation = Animated.add(el.a, el.parent.rotation);
                const TETA = Animated.multiply(totalRotation, Math.PI / 180);

                const COS = TETA.interpolate(WithFunction(Math.cos, 20, 2*Math.PI));
                const SIN = TETA.interpolate(WithFunction(Math.sin, 20, 2*Math.PI));
                const radiusX = Animated.multiply(this.l, COS);
                const radiusY = Animated.multiply(this.l, SIN);

                const X = Animated.add(el.parent.pos.x, radiusX);
                const Y = Animated.add(el.parent.pos.y, radiusY);

                const parentRotation = Animated.add(el.parent.a, el.parent.rotation);
                R = Animated.add(parentRotation, el.r);
                return [ X, Y, R ];

                calculateParentPos(el.parent);
            }
        }*/



        let posX = this.pos.x;
        let posY = this.pos.y;
        let rotZ = this.rotation;

        /** @param {Part} el */
        const calculateParentPos = (el) => {
            if (el.parent !== null) {
                const parentRotation = Animated.modulo(Animated.add(el.parent.rotation, el.parent.a), 360);
                //const totalRotation = Animated.modulo(Animated.add(parentRotation, el.a), 360);
                const totalRotation = Animated.modulo(Animated.add(el.parent.rotation, el.a), 360);

                let _totalRotation = new Animated.Value(0);
                const elChilds = el.getParents();
                for (let i = 0; i < elChilds.length-1; i++) {
                    const child = elChilds[i];
                    const childRotation = Animated.add(child.a % 360, child.rotation);
                    _totalRotation = Animated.add(_totalRotation, childRotation);
                }
                _totalRotation = Animated.add(_totalRotation, el.a);
                _totalRotation = Animated.modulo(_totalRotation, 360);

                const TETA = Animated.multiply(_totalRotation, Math.PI / 180);
                const COS = TETA.interpolate(WithFunction(Math.cos, 20, 2*Math.PI));
                const SIN = TETA.interpolate(WithFunction(Math.sin, 20, 2*Math.PI));
                const radiusX = Animated.multiply(el.l, COS);
                const radiusY = Animated.multiply(el.l, SIN);

                const parentX = Animated.add(el.parent.pos.x, radiusX);
                const parentY = Animated.add(el.parent.pos.y, radiusY);

                posX = Animated.add(posX, parentX);
                posY = Animated.add(posY, parentY);
                rotZ = Animated.add(rotZ, parentRotation);

                calculateParentPos(el.parent);
            }
        }
        calculateParentPos(this);

        const character = (
            <Character_Male_Test
                key={'part-' + this.name + '-' + Math.random()}
                part={this.name}
                fill='#e0a98b'
                posX={posX}
                posY={posY}
                rotation={rotZ}
                zIndex={this.zIndex}
            />
        );

        if (this.name !== 'bust') return character;

        /** @param {Part} part */
        const allRender = (part) => part == this ? character : part.render();
        // TODO - Now: show only this and childs, but in future: show all parts from bust
        const components = [this, ...this.childs].sort((a, b) => a.zIndex - b.zIndex).map(allRender);
        return (components);
    }
}

const CharacterProps = {
}

class Character extends React.Component {
    state = {
        positions: {
            bust: new Animated.ValueXY({ x: 0, y: 0 })
        },
        rotations: {
            bust: new Animated.Value(0),
            head: new Animated.Value(0),
            left_arm: new Animated.Value(0),
            left_forearm: new Animated.Value(-130),
            left_hand: new Animated.Value(20),
            right_arm: new Animated.Value(0),
            right_forearm: new Animated.Value(-20),
            right_hand: new Animated.Value(-20),
            left_thigh: new Animated.Value(0),
            left_leg: new Animated.Value(0),
            left_foot: new Animated.Value(0),
            right_thigh: new Animated.Value(0),
            right_leg: new Animated.Value(0),
            right_foot: new Animated.Value(0)
        }
    }

    constructor(props) {
        super(props);

        const pos = this.state.positions;
        const rot = this.state.rotations;
        this.pos = { x: 450, y: 280 };
        pos.bust.addListener(this.__setPosition.bind(this));

        this.body = new Part('bust', 0, 0, rot.bust, 1);
        this.body.SetPosition(pos.bust);
        const head = new Part('head', 0, 0, rot.head, 2);

        const left_arm = new Part('left_arm', 100, 135, rot.left_arm, 2);
        const left_forearm = new Part('left_forearm', 120, -25, rot.left_forearm, 2);
        const left_hand = new Part('left_hand', 100, 105, rot.left_hand, 5);

        const right_arm = new Part('right_arm', 110, 30, rot.right_arm, -1);
        const right_forearm = new Part('right_forearm', 140, 40, rot.right_forearm);
        const right_hand = new Part('right_hand', 90, 25, rot.right_hand);

        const left_thigh = new Part('left_thigh', 0, 0, rot.left_thigh);
        const left_leg = new Part('left_leg', 0, 0, rot.left_leg);
        const left_foot = new Part('left_foot', 0, 0, rot.left_foot);
        const right_thigh = new Part('right_thigh', 0, 0, rot.right_thigh);
        const right_leg = new Part('right_leg', 0, 0, rot.right_leg);
        const right_foot = new Part('right_foot', 0, 0, rot.right_foot);

        this.body.AddChild(head);
        // Left arm
        left_forearm.AddChild(left_hand);
        left_arm.AddChild(left_forearm);
        this.body.AddChild(left_arm);
        // Right arm
        right_forearm.AddChild(right_hand);
        right_arm.AddChild(right_forearm);
        this.body.AddChild(right_arm);
        // Left leg
        //left_leg.AddChild(left_foot);
        //left_thigh.AddChild(left_leg);
        //this.body.AddChild(left_thigh);
        // Right leg
        //right_leg.AddChild(right_foot);
        //right_thigh.AddChild(right_leg);
        //this.body.AddChild(right_thigh);

        //this.character = this.body.render();
    }

    componentDidMount() {
        //const angles = [ 0, 90, -10, 0, -20, 40, 20, -25, -30, 0, -20, -40, 0 ];
        //const angles = [ 20, 40, 0, -20, -40, 40, 60, -25, -30, 0, -20, -40, 0 ];
        this.SetPartRotation('head', 10, 1000);
        this.SetPositionAbsolute(this.pos.x, this.pos.y, 0);

        const ToIdle = () => {
            this.SetPartRotation('bust', 0, 1000);
            this.SetPartRotation('head', 0, 1000);
            this.SetPartRotation('left_arm', 0, 1000);
            this.SetPartRotation('left_forearm', -130, 1000);
            this.SetPartRotation('left_hand', 20, 1000);
            this.SetPartRotation('right_arm', 0, 1000);
            this.SetPartRotation('right_forearm', -20, 1000);
            this.SetPartRotation('right_hand', -20, 1000);
            this.SetPartRotation('left_thigh', 0, 1000);
            this.SetPartRotation('left_leg', 0, 1000);
            this.SetPartRotation('left_foot', 0, 1000);
            this.SetPartRotation('right_thigh', 0, 1000);
            this.SetPartRotation('right_leg', 0, 1000);
            this.SetPartRotation('right_foot', 0, 1000);
        }

        const ToMuscles = () => {
            this.SetPartRotation('bust', 0, 1000);
            this.SetPartRotation('head', 20, 1000);
            this.SetPartRotation('left_arm', 40, 1000);
            this.SetPartRotation('left_forearm', -200, 1000);
            this.SetPartRotation('left_hand', -20, 1000);
            this.SetPartRotation('right_arm', 0, 1000);
            this.SetPartRotation('right_forearm', -20, 1000);
            this.SetPartRotation('right_hand', -20, 1000);
            this.SetPartRotation('left_thigh', 0, 1000);
            this.SetPartRotation('left_leg', 0, 1000);
            this.SetPartRotation('left_foot', 0, 1000);
            this.SetPartRotation('right_thigh', 0, 1000);
            this.SetPartRotation('right_leg', 0, 1000);
            this.SetPartRotation('right_foot', 0, 1000);
        }

        //setTimeout(ToIdle, 1000);
        setTimeout(ToMuscles, 100);

        return;
        setTimeout(() => this.SetPartRotation('left_arm', 400, 750), 1000);
        setTimeout(() => this.SetPartRotation('left_forearm', -20, 750), 2000);
        setTimeout(() => this.SetPartRotation('bust', -20, 750), 3000);
        //setTimeout(() => this.SetPartRotation('left_hand', -20, 750), 4000);
        //setTimeout(() => this.SetPositionRelative(150, 150), 4000);
    }

    SetPositionAbsolute(x, y, duration = 1000) {
        TimingAnimation(this.state.positions.bust, { x, y }, duration).start();
    }
    SetPositionRelative(x, y, duration = 1000) {
        const newX = this.pos.x + x;
        const newY = this.pos.y + y;
        TimingAnimation(this.state.positions.bust, { x: newX, y: newY }, duration).start();
    }
    __setPosition({ x, y }) {
        this.pos = { x, y };
    }

    /**
     * @param {'bust'|'head'|'left_arm'|'left_forearm'|'left_hand'|'right_arm'|'right_forearm'|'right_hand'|'left_thigh'|'left_leg'|'left_foot'|'right_thigh'|'right_leg'|'right_foot'} partName
     * @param {Number} angle Rotations of body part
     */
    SetPartRotation = (partName, angle = 0, duration = 1000) => {
        if (!Object.keys(this.state.rotations).includes(partName)) {
            console.warn('Character: Wrong part name');
            return;
        }
        TimingAnimation(this.state.rotations[partName], angle, duration).start();
    }

    render() {
        return (
            <View style={styles.canvas}>
                <Svg viewBox='0 0 1000 1000'>
                    {this.body.render()}
                </Svg>
            </View>
        );
    }
}

Character.prototype.props = CharacterProps;
Character.defaultProps = CharacterProps;

const styles = StyleSheet.create({
    canvas: {
        width: '100%',
        height: '100%'
    },
    part: {
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%'
    }
});

export default Character;