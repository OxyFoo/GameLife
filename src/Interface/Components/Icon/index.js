import * as React from 'react';
import { View, TouchableOpacity } from 'react-native';
import { SvgXml } from 'react-native-svg';

import SVGIcons from './icons';
import themeManager from 'Managers/ThemeManager';

import Base64 from 'Utils/Base64';

/**
 * @typedef {import('react-native').ViewStyle} ViewStyle
 * @typedef {import('react-native').StyleProp<ViewStyle>} StyleProp
 * @typedef {import('react-native').GestureResponderEvent} GestureResponderEvent
 * 
 * @typedef {import('Managers/ThemeManager').ColorTheme} ColorTheme
 * @typedef {import('Managers/ThemeManager').ColorThemeText} ColorThemeText
 * @typedef {'default'|'add'|'alarmClock'|'arrowLeft'|'calendar'|'check'|'checkboxOn'|'checkboxOff'|'chevron'|'chrono'|'cross'|'discord'|'edit'|'error'|'filter'|'flagEnglish'|'flagFrench'|'home'|'human'|'info'|'instagram'|'item'|'moveVertical'|'nowifi'|'onboarding1'|'onboarding2'|'onboarding3'|'ox'|'setting'|'shop'|'sleepZzz'|'social'|'success'|'tiktok'|'userAdd'|'world'|'loading'|'loadingDots'} Icons
 *
 * @callback GestureEvent
 * @param {GestureResponderEvent} event
 */

const IconProps = {
    /** @type {StyleProp} */
    style: {},

    /** @type {StyleProp} */
    containerStyle: {},

    /** @type {string|null} Display an icon from XML base64 encoded ('icon' skip if define) */
    xml: null,

    /** @type {Icons|null} */
    icon: null,

    /** @type {number} Size of icon in pixels */
    size: 24,

    /** @type {number} Rotation angle in degrees */
    angle: 0,

    /** @type {ColorTheme|ColorThemeText} */
    color: 'white',

    /** @type {(event: GestureResponderEvent) => void|null} */
    onPress: null,

    /** @type {boolean} */
    show: true
}

class Icon extends React.Component {
    render() {
        let output;
        const { style, containerStyle, icon, xml, size, angle, onPress, show } = this.props;
        const containerSize = { width: size, height: size };
        const color = themeManager.GetColor(this.props.color);

        // Icon
        if (show && xml !== null) {
            // Default icon (invalid xml)
            if (typeof(xml) !== 'string' || xml.length < 10) {
                output = <Icon icon='default' size={size} color={this.props.color} />;
            }

            // XML icon
            else {
                const XML = Base64.Decode(xml)
                    .split('#ffffff').join(color)
                    .split('#FFFFFF').join(color);
                output = <View style={[containerSize, style]}>
                            <SvgXml xml={XML} width={size} height={size} />
                        </View>;
            }
        }

        // Icon
        else if (show && icon !== null && SVGIcons.hasOwnProperty(icon)) {
            const _Icon = SVGIcons[icon];
            output = <View style={[containerSize, style]}>
                        <_Icon
                            width={size}
                            height={size}
                            color={color}
                            transform={[{ rotate: angle * Math.PI / 180 }]}
                        />
                    </View>;
        }

        // Empty icon
        else {
            output = <View style={[containerSize, style]} />;
        }

        if (onPress !== null) {
            output = <TouchableOpacity
                        testID='icon'
                        style={[containerSize, containerStyle]}
                        onPress={onPress}
                        activeOpacity={.5}
                    >
                        {output}
                    </TouchableOpacity>;
        }

        return output;
    }
}

Icon.prototype.props = IconProps;
Icon.defaultProps = IconProps;

export default Icon;