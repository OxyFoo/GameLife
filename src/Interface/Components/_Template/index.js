import * as React from 'react';
import { View } from 'react-native';

import styles from './style';

class NewComponent extends React.Component {
    render() {
        return (
            <View>
            </View>
        );
    }
}

export default NewComponent;
