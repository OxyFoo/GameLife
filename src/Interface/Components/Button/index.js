import * as React from 'react';
import { View, Animated } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import { BlurView } from '@react-native-community/blur';
import MaskedView from '@react-native-masked-view/masked-view';

import styles from './style';
import ButtonBack from './back';
import themeManager from 'Managers/ThemeManager';

import { Text } from '../Text';
import { Icon } from '../Icon';
import { Ripple } from 'Interface/Primitives';

/**
 * @typedef {import('react-native').ViewStyle} ViewStyle
 * @typedef {import('react-native').StyleProp<ViewStyle>} StyleProp
 */

class Button extends ButtonBack {
    render() {
        const {
            appearance,
            style: styleProp,
            styleAnimation,
            styleContent,
            enabled,
            onTouchStart,
            onTouchCancel,
            onTouchMove,
            onTouchEnd,
            onLayout,
            children,
            icon,
            iconSize,
            iconXml,
            iconAngle,
            loading,
            fontSize,
            color,
            pointerEvents,
            onPress,
            onLongPress,
            ...rest
        } = this.props;

        const ButtonView = styleAnimation === null ? View : Animated.View;
        return (
            // @ts-ignore
            <ButtonView
                {...rest}
                style={[styles.body, styleProp, styleAnimation]}
                onTouchStart={this.onTouchStart}
                onTouchCancel={this.onTouchCancel}
                onTouchMove={this.onTouchMove}
                onTouchEnd={this.onTouchEnd}
                onLayout={this.onLayout}
                accessible={true}
                accessibilityRole='button'
                accessibilityState={{ disabled: !enabled }}
                pointerEvents={pointerEvents}
            >
                {this.renderBackground()}
                {this.renderContent()}

                {/** Ripple */}
                {enabled && <Ripple ref={this.rippleRef} rippleColor={appearance === 'normal' ? 'black' : 'white'} />}
            </ButtonView>
        );
    }

    renderContent = () => {
        const {
            children,
            appearance,
            loading,
            icon,
            iconXml,
            styleContent: styleContentProp,
            color,
            fontSize,
            fontColor,
            enabled
        } = this.props;

        const hasChildren = typeof children !== 'undefined';
        const hasIcon = icon !== null || iconXml !== null;

        let content = children;
        let childCount = 1;

        let _fontColor = fontColor;
        if (_fontColor === 'automatic') {
            if (appearance === 'uniform') {
                const luminance = themeManager.GetLuminance(themeManager.GetColor(color));
                _fontColor = luminance > 0.75 ? 'darkBlue' : 'white';
            } else {
                _fontColor = 'darkBlue';
            }
        }

        // Loading icon
        if (loading) {
            content = (
                <Icon
                    style={styles.loadingIcon}
                    icon='loading-dots'
                    size={this.props.iconSize - 2}
                    color={_fontColor}
                />
            );
        }

        // Manage children
        else if (hasChildren) {
            if (typeof children === 'string') {
                content = (
                    <Text color={_fontColor} fontSize={fontSize}>
                        {children}
                    </Text>
                );
            }

            // Add icon after text
            if (hasIcon) {
                childCount = 2;
                content = (
                    <View style={styles.content}>
                        <Icon icon={'default'} size={this.props.iconSize} color={'transparent'} />
                        <View style={[styles.content, styles.flex]}>{content}</View>
                        <Icon
                            icon={this.props.icon}
                            xml={this.props.iconXml}
                            size={this.props.iconSize}
                            color={_fontColor}
                            angle={this.props.iconAngle}
                        />
                    </View>
                );
            }
        }

        // Icon only
        else if (!hasChildren && hasIcon) {
            content = (
                <Icon
                    icon={this.props.icon}
                    xml={this.props.iconXml}
                    size={this.props.iconSize}
                    color={_fontColor}
                    angle={this.props.iconAngle}
                />
            );
        }

        /** @type {StyleProp} */
        const styleContent = {
            justifyContent: childCount === 1 ? 'center' : 'space-between',
            opacity: enabled ? 1 : 0.6
        };

        if (appearance === 'normal' || appearance === 'uniform' || fontColor !== 'automatic') {
            return (
                <View style={[styles.content, styleContent, styleContentProp]} pointerEvents='none'>
                    {content}
                </View>
            );
        } else if (appearance === 'outline' || appearance === 'outline-blur') {
            return (
                <View>
                    <View style={[styles.content, styleContent, styleContentProp]} pointerEvents='none'>
                        {content}
                    </View>
                    <MaskedView
                        style={styles.absolute}
                        maskElement={<View style={[styles.content, styleContent, styleContentProp]}>{content}</View>}
                    >
                        <LinearGradient
                            style={styles.fill}
                            colors={['#8CF7FF', '#DBA1FF']}
                            useAngle={true}
                            angle={190}
                        />
                    </MaskedView>
                </View>
            );
        }

        return null;
    };

    renderBackground = () => {
        const { appearance, color } = this.props;

        if (appearance === 'normal') {
            return (
                <LinearGradient style={styles.absolute} colors={['#8CF7FF', '#DBA1FF']} useAngle={true} angle={267} />
            );
        } else if (appearance === 'uniform') {
            return (
                <View
                    style={[
                        styles.absolute,
                        appearance === 'uniform' && {
                            backgroundColor: themeManager.GetColor(color)
                        }
                    ]}
                />
            );
        } else if (appearance === 'outline') {
            return (
                <MaskedView style={styles.absolute} maskElement={<View style={styles.backgroundView} />}>
                    <LinearGradient style={styles.fill} colors={['#8CF7FF', '#DBA1FF']} useAngle={true} angle={267} />
                </MaskedView>
            );
        } else if (appearance === 'outline-blur') {
            return (
                <>
                    <BlurView
                        style={[
                            styles.absolute,
                            styles.backgroundBlur,
                            {
                                backgroundColor: themeManager.GetColor('darkBlue', {
                                    opacity: 0.25
                                })
                            }
                        ]}
                        blurAmount={20}
                    />
                    <MaskedView style={styles.absolute} maskElement={<View style={styles.backgroundView} />}>
                        <LinearGradient
                            style={styles.fill}
                            colors={['#8CF7FF', '#DBA1FF']}
                            useAngle={true}
                            angle={267}
                        />
                    </MaskedView>
                </>
            );
        }

        return null;
    };
}

export { Button };
