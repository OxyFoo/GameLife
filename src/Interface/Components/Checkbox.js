import * as React from 'react';
import { View, Animated, StyleSheet } from 'react-native';

import themeManager from '../../Managers/ThemeManager';

import Icon from './Icon';
import { SpringAnimation } from '../../Utils/Animations';

const CheckboxProps = {
    color: 'main1',
    style: {},
    checked: false,
    onChange: undefined
}

class Checkbox extends React.Component {
    state = {
        animScale: new Animated.Value(0)
    }

    componentDidUpdate() {
        const newValue = this.props.checked ? 1 : 0;
        SpringAnimation(this.state.animScale, newValue).start();
    }

    onPress = () => {
        if (typeof(this.props.onChange) === 'function') {
            this.props.onChange();
        }
    }

    render() {
        const activeColor = themeManager.GetColor(this.props.color);
        const emptyColor = themeManager.GetColor('border');

        return (
            <View style={this.props.style}>
                <Icon onPress={this.onPress} icon={'checkboxOff'} color={emptyColor} />
                <Animated.View style={[styles.check, { transform: [{ scale: this.state.animScale }] }]} pointerEvents={'none'}>
                    <Icon icon={'checkboxOn'} color={activeColor}></Icon>
                </Animated.View>
            </View>
        );
    }
}

Checkbox.prototype.props = CheckboxProps;
Checkbox.defaultProps = CheckboxProps;

const styles = StyleSheet.create({
    check: {
        position: 'absolute',
        top: 0,
        left: 0
    }
});

export default Checkbox;