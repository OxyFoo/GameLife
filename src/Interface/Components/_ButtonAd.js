import * as React from 'react';
import { View, StyleSheet } from 'react-native';

import user from '../../Managers/UserManager';
import langManager from '../../Managers/LangManager';

import Text from './Text';
import Icon from './Icon';

const ButtonAdProps = {
    button: null,
    oxAmount: 10,

    /** @type {Boolean?} True if ad is free */
    free: false
}

class ButtonAd extends React.PureComponent {
    render() {
        const Button = this.props.button;
        if (Button === null) {
            user.interface.console.AddLog('warn', 'ButtonAd: button prop is null');
            return null;
        }

        const { oxAmount, free } = this.props;
        const lang = langManager.curr['other'];
        const enabled = free && user.informations.adRemaining > 0;
        const style = [styles.adButton, this.props.style];
        const text = lang['ad-button-text'];

        let error = null;
        if (free === false) error = lang['ad-button-empty'];
        else if (free === null) error = lang['ad-button-error'];

        return (
            <Button color='main2' {...this.props} style={style} enabled={enabled}>
                <Text>{error || text}</Text>
                {!error && (
                    <View style={styles.adIcon}>
                        <Text style={styles.adText}>+{oxAmount}</Text>
                        <Icon icon='ox' color='white' size={24} />
                    </View>
                )}
            </Button>
        );
    }
}

ButtonAd.prototype.props = ButtonAdProps;
ButtonAd.defaultProps = ButtonAdProps;

const styles = StyleSheet.create({
    adButton: { justifyContent: 'space-between', borderRadius: 14 },
    adText: { marginRight: 6 },
    adIcon: { flexDirection: 'row' }
});

export { ButtonAdProps };
export default ButtonAd;