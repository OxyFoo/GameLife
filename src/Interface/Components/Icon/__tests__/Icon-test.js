import 'react-native';
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

// Graphical components
import { Icon } from 'Interface/Components'

describe('[Component] Icon', () => {
    it('renders correctly', () => {
        const { toJSON } = render(<Icon />);
        expect(toJSON()).toMatchSnapshot();
    });

    it('should call a callback function on icon click', () => {
        const onClickMock = jest.fn();
        render(<Icon onPress={onClickMock} />);

        const iconElement = render(<Icon onPress={onClickMock} />).getByTestId('icon');
        fireEvent.press(iconElement);
        expect(onClickMock).toHaveBeenCalled();
    });
});
