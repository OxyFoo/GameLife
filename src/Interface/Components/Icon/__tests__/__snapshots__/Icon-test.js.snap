// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Icon renders correctly 1`] = `
<View
  style={
    [
      {
        "height": 24,
        "width": 24,
      },
      {},
    ]
  }
/>
`;
